File: src/motor-model.cpp
/*
 * Copyright 2014, Andrea Del Prete, LAAS-CNRS
 *
 */

#include <cassert>
#include <cmath>
#include <sot/torque_control/motor-model.hh>
/// i = Kt*tau + Kv*dq + Ka*ddq + sign(dq)Kf
/// tau = i/Kt - (Kv/Kt)*dq - (Ka/Kt)*ddq - sign(dq)(Kf/Kt)
namespace dynamicgraph {
namespace sot {
namespace torque_control {
MotorModel::MotorModel() {}

double MotorModel::getCurrent(double torque, double dq, double ddq, double Kt_p,
                              double Kt_n, double Kf_p, double Kf_n,
                              double Kv_p, double Kv_n, double Ka_p,
                              double Ka_n, unsigned int poly) {
  assert(Kt_p > 0.0 && "Kt_p should be > 0");
  assert(Kt_n > 0.0 && "Kt_n should be > 0");
  assert(Kf_p >= 0.0 && "Kf_p should be >= 0");
  assert(Kf_n >= 0.0 && "Kf_n should be >= 0");
  assert(Kv_p >= 0.0 && "Kv_p should be >= 0");
  assert(Kv_n >= 0.0 && "Kv_n should be >= 0");
  assert(Ka_p >= 0.0 && "Ka_p should be >= 0");
  assert(Ka_n >= 0.0 && "Ka_n should be >= 0");

  double signDq = this->smoothSign(dq, 0.1, poly);  // in [-1;1]
  double current;

  // Smoothly set Coefficients according to velocity sign
  double Kt = 0.5 * (Kt_p * (1 + signDq) + Kt_n * (1 - signDq));
  double Kv = 0.5 * (Kv_p * (1 + signDq) + Kv_n * (1 - signDq));
  double Ka = 0.5 * (Ka_p * (1 + signDq) + Ka_n * (1 - signDq));
  double Kf = 0.5 * (Kf_p * (1 + signDq) + Kf_n * (1 - signDq));

  current = Kt * torque + Kv * dq + Ka * ddq + signDq * Kf;

  return current;
}

double MotorModel::getTorque(double current, double dq, double ddq, double Kt_p,
                             double Kt_n, double Kf_p, double Kf_n, double Kv_p,
                             double Kv_n, double Ka_p, double Ka_n,
                             unsigned int poly) {
  assert(Kt_p > 0.0 && "Kt_p should be > 0");
  assert(Kt_n > 0.0 && "Kt_n should be > 0");
  assert(Kf_p >= 0.0 && "Kf_p should be >= 0");
  assert(Kf_n >= 0.0 && "Kf_n should be >= 0");
  assert(Kv_p >= 0.0 && "Kv_p should be >= 0");
  assert(Kv_n >= 0.0 && "Kv_n should be >= 0");
  assert(Ka_p >= 0.0 && "Ka_p should be >= 0");
  assert(Ka_n >= 0.0 && "Ka_n should be >= 0");

  double signDq = this->smoothSign(dq, 0.1, poly);  // in [-1;1]
  double torque;

  // Smoothly set Coefficients according to velocity sign
  double Kt = 0.5 * (Kt_p * (1 + signDq) + Kt_n * (1 - signDq));
  double Kv = 0.5 * (Kv_p * (1 + signDq) + Kv_n * (1 - signDq));
  double Ka = 0.5 * (Ka_p * (1 + signDq) + Ka_n * (1 - signDq));
  double Kf = 0.5 * (Kf_p * (1 + signDq) + Kf_n * (1 - signDq));

  torque =
      (current / Kt) - (Kv / Kt) * dq - (Ka / Kt) * ddq - signDq * (Kf / Kt);

  return torque;
}

double MotorModel::smoothSign(double value, double threshold,
                              unsigned int poly) {
  if (value > threshold)
    return 1.0;
  else if (value < -threshold)
    return -1.0;
  double a = value / threshold;
  if (poly == 1) return a;
  if (poly == 2 && value > 0) return a * a;
  if (poly == 2 && value <= 0) return -a * a;
  return a * a * a;
}
}  // namespace torque_control
}  // namespace sot
}  // namespace dynamicgraph

File: python/dynamic_graph/sot/torque_control/identification/motor_model.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# i = Kt*tau + Kv*dq + Ka*ddq + sign(dq)Kf
# tau = i/Kt - (Kv/Kt)*dq - (Ka/Kt)*ddq - sign(dq)(Kf/Kt)
from IPython import embed  # noqa


def smoothSign(value, threshold):
    if value > threshold:
        return 1.0
    elif value < -threshold:
        return -1.0
    return pow(value / threshold, 3)


class Motor_model:
    def __init__(self, Kt_p, Kt_n, Kf_p, Kf_n, Kv_p, Kv_n, Ka_p, Ka_n, dqThreshold=0.8):
        self.Kt_p = Kt_p
        self.Kt_n = Kt_n

        self.Kf_p = Kf_p
        self.Kf_n = Kf_n

        self.Kv_p = Kv_p
        self.Kv_n = Kv_n

        self.Ka_p = Ka_p
        self.Ka_n = Ka_n
        self.dqThreshold = dqThreshold

    def getCurrent(self, tau, dq, ddq):
        Kt_p = self.Kt_p
        Kt_n = self.Kt_n

        Kf_p = self.Kf_p
        Kf_n = self.Kf_n

        Kv_p = self.Kv_p
        Kv_n = self.Kv_n

        Ka_p = self.Ka_p
        Ka_n = self.Ka_n

        signDq = smoothSign(dq, self.dqThreshold)
        # in [-1;1]
        # Smoothly set Coefficients according to velocity sign
        Kt = 0.5 * (Kt_p * (1 + signDq) + Kt_n * (1 - signDq))
        Kv = 0.5 * (Kv_p * (1 + signDq) + Kv_n * (1 - signDq))
        Ka = 0.5 * (Ka_p * (1 + signDq) + Ka_n * (1 - signDq))
        Kf = 0.5 * (Kf_p * (1 + signDq) + Kf_n * (1 - signDq))
        current = Kt * tau + Kv * dq + Ka * ddq + signDq * Kf
        return current

    def getTorque(self, current, dq, ddq):
        Kt_p = self.Kt_p
        Kt_n = self.Kt_n

        Kf_p = self.Kf_p
        Kf_n = self.Kf_n

        Kv_p = self.Kv_p
        Kv_n = self.Kv_n

        Ka_p = self.Ka_p
        Ka_n = self.Ka_n

        signDq = smoothSign(dq, self.dqThreshold)
        # in [-1;1]
        # Smoothly set Coefficients according to velocity sign
        Kt = 0.5 * (Kt_p * (1 + signDq) + Kt_n * (1 - signDq))
        Kv = 0.5 * (Kv_p * (1 + signDq) + Kv_n * (1 - signDq))
        Ka = 0.5 * (Ka_p * (1 + signDq) + Ka_n * (1 - signDq))
        Kf = 0.5 * (Kf_p * (1 + signDq) + Kf_n * (1 - signDq))
        torque = (current / Kt) - (Kv / Kt) * dq - (Ka / Kt) * ddq - signDq * (Kf / Kt)
        return torque
